### PR TITLE
feat: allow users to bring their own tonic channel.

### LINKED ISSUE #486: opentelemetry_otlp not support uds endpoint
use unix domain socket address get err : ExportFailed(InvalidUri(InvalidUri(InvalidFormat)))

### PR DESCRIPTION
**Pull request by TommyCpp**
Sometimes users want to use custom channel like UDS. Added a `from_tonic_channel` function to allow users to bring the custom channel.

fixes #486

### CODE REVIEW

**Comment by jtescher** [comment]
Seems like a good idea to allow users to pass a channel when using tonic, could this be added through the builder to avoid the need to synchronize timeouts and generally simplify things?

**Comment by jtescher** [comment]
Hm guess no reader so maybe not simpler

**Comment by TommyCpp** [comment]
> could this be added through the builder to avoid the need to synchronize timeouts and generally simplify things

Adding it to the builder may simplify things as users won't need to manually construct config structs. 

But users will still need to sync the timeouts because we need to set up the timeout in [`Endpoint`](https://docs.rs/tonic/0.4.1/tonic/transport/struct.Endpoint.html#method.timeout)s instead of [`Channel`](https://docs.rs/tonic/0.4.1/tonic/transport/struct.Channel.html)s. Thus we cannot change the timeout of the user-provided channel.

**Comment by TommyCpp** [comment]
Furthermore, we may want to find a way to allow users to provide more configurations in `Endpoint` like [concurrent_limit](https://docs.rs/tonic/0.4.1/tonic/transport/struct.Endpoint.html#method.concurrency_limit), [rate_limit](https://docs.rs/tonic/0.4.1/tonic/transport/struct.Endpoint.html#method.rate_limit)

File: opentelemetry-otlp/src/span.rs [hunk 1]
```diff
@@ -206,16 +206,16 @@ impl Debug for TraceExporter {
...
     pub fn new_tonic(
         config: ExporterConfig,
         tonic_config: TonicConfig,
     ) -> Result<Self, crate::Error> {
-        let endpoint = TonicChannel::from_shared(config.endpoint)?;
+        let endpoint = TonicChannel::from_shared(config.endpoint.clone())?;
 
         #[cfg(feature = "tls")]
-        let channel = match tonic_config.tls_config {
+        let channel = match tonic_config.tls_config.clone() {
```
Comment (djc) [inline opentelemetry-otlp/src/span.rs:218]: I think we could only clone here in the `Some` case?

**Comment by djc** [comment]
The design constraint that the timeouts must match feel like something that should usually be expressed in the type system somehow rather than only in the documentation. Maybe the additional constructor should take an `Endpoint` rather than a `Channel` (and document that it overwrites the timeout)?

**Comment by TommyCpp** [comment]
> Maybe the additional constructor should take an `Endpoint` rather than a `Channel` (and document that it overwrites the timeout)

Sometimes ppl need to pass a connector function into `Endpoint` like in  [Uds](https://github.com/hyperium/tonic/blob/master/examples/src/uds/client.rs#L21-L26) example. Thus we might still need them to pass the constructed channel.

I also thought about allowing users to pass the connector function but the type of it seems to be bind with `tower` and is not trivial.

**Review by jtescher — APPROVED** [review]
Looks reasonable to me, @djc did you have other comments?

File: opentelemetry-otlp/src/lib.rs [hunk 2]
```diff
@@ -344,6 +346,19 @@ impl TonicPipelineBuilder {
         self
     }
 
+    /// Use `channel` as tonic's transport channel.
+    /// this will override tls config and should only be used
+    /// when working with non-HTTP transport.
```
Comment (djc) [inline opentelemetry-otlp/src/lib.rs:351]: Nit: I think `transport` should be plural here?

**Review by djc — APPROVED** [review]
Just one nit.

File: opentelemetry-otlp/src/lib.rs (hunk 2 above)
Comment (jtescher) [reply opentelemetry-otlp/src/lib.rs:351]: @TommyCpp I think he meant the other one, first line singular, last line plural

Comment (TommyCpp) [reply opentelemetry-otlp/src/lib.rs]: Oops sorry I misunderstand that one 😬

**Pull request merged**